File: FriendlyTerminal/Models/ProjectCommands.swift
import Foundation

struct ProjectCommandGroup: Identifiable {
    let name: String
    let systemImage: String
    let commands: [ProjectCommand]
    var id: String { name }
}

struct ProjectCommand: Identifiable {
    let label: String
    let command: String
    let systemImage: String
    var id: String { command }
}

enum ProjectCommandDetector {
    static func suggestions(cwd: String, fileItems: [FileItem]) -> [ProjectCommandGroup] {
        let names = Set(fileItems.map(\.name))
        var groups: [ProjectCommandGroup] = []

        if names.contains("package.json") {
            if let group = nodeGroup(cwd: cwd, names: names) {
                groups.append(group)
            }
        }

        if names.contains("requirements.txt") || names.contains("pyproject.toml") || names.contains("setup.py") {
            groups.append(pythonGroup(cwd: cwd, names: names))
        }

        if names.contains("Cargo.toml") {
            groups.append(ProjectCommandGroup(
                name: "Rust", systemImage: "gearshape",
                commands: [
                    ProjectCommand(label: "Run", command: "cargo run", systemImage: "play.fill"),
                    ProjectCommand(label: "Build", command: "cargo build", systemImage: "hammer.fill"),
                    ProjectCommand(label: "Test", command: "cargo test", systemImage: "checkmark.seal.fill"),
                ]
            ))
        }

        if names.contains("go.mod") {
            groups.append(ProjectCommandGroup(
                name: "Go", systemImage: "chevron.left.forwardslash.chevron.right",
                commands: [
                    ProjectCommand(label: "Run", command: "go run .", systemImage: "play.fill"),
                    ProjectCommand(label: "Test", command: "go test ./...", systemImage: "checkmark.seal.fill"),
                    ProjectCommand(label: "Build", command: "go build .", systemImage: "hammer.fill"),
                ]
            ))
        }

        if names.contains("Makefile") {
            groups.append(ProjectCommandGroup(
                name: "Make", systemImage: "hammer",
                commands: makeTargets(cwd: cwd)
            ))
        }

        if names.contains("Gemfile") {
            groups.append(ProjectCommandGroup(
                name: "Ruby", systemImage: "diamond",
                commands: [
                    ProjectCommand(label: "Bundle install", command: "bundle install", systemImage: "square.and.arrow.down"),
                    ProjectCommand(label: "RSpec", command: "bundle exec rspec", systemImage: "checkmark.seal.fill"),
                ]
            ))
        }

        if names.contains("Dockerfile") {
            groups.append(ProjectCommandGroup(
                name: "Docker", systemImage: "shippingbox",
                commands: [
                    ProjectCommand(label: "Build", command: "docker build .", systemImage: "hammer.fill"),
                ]
            ))
        }

        return groups
    }

    private static func nodeGroup(cwd: String, names: Set<String>) -> ProjectCommandGroup? {
        let manager: String
        if names.contains("pnpm-lock.yaml") || names.contains("pnpm-lock.yml") {
            manager = "pnpm"
        } else if names.contains("yarn.lock") {
            manager = "yarn"
        } else {
            manager = "npm"
        }

        var commands: [ProjectCommand] = []

        if !names.contains("node_modules") {
            let installCmd = manager == "npm" ? "npm install" : "\(manager) install"
            commands.append(ProjectCommand(
                label: "Install deps", command: installCmd, systemImage: "square.and.arrow.down"
            ))
        }

        let pkgPath = (cwd as NSString).appendingPathComponent("package.json")
        if let data = try? Data(contentsOf: URL(fileURLWithPath: pkgPath)),
           let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let scripts = json["scripts"] as? [String: Any] {

            let priority = ["dev", "start", "serve", "preview", "build", "test", "lint", "check", "format"]
            for key in priority where scripts[key] != nil {
                let cmd = manager == "npm" ? "npm run \(key)" : "\(manager) \(key)"
                commands.append(ProjectCommand(
                    label: key, command: cmd, systemImage: scriptIcon(key)
                ))
            }
            let extra = scripts.keys.filter { !priority.contains($0) }.sorted().prefix(4)
            for key in extra {
                let cmd = manager == "npm" ? "npm run \(key)" : "\(manager) \(key)"
                commands.append(ProjectCommand(label: key, command: cmd, systemImage: "terminal"))
            }
        } else {
            let runDev = manager == "npm" ? "npm run dev" : "\(manager) dev"
            let runTest = manager == "npm" ? "npm test" : "\(manager) test"
            let runBuild = manager == "npm" ? "npm run build" : "\(manager) build"
            commands += [
                ProjectCommand(label: "dev", command: runDev, systemImage: "play.fill"),
                ProjectCommand(label: "test", command: runTest, systemImage: "checkmark.seal.fill"),
                ProjectCommand(label: "build", command: runBuild, systemImage: "hammer.fill"),
            ]
        }

        guard !commands.isEmpty else { return nil }
        let displayName: String
        switch manager {
        case "pnpm": displayName = "pnpm"
        case "yarn": displayName = "Yarn"
        default:     displayName = "Node.js"
        }
        return ProjectCommandGroup(name: displayName, systemImage: "curlybraces", commands: commands)
    }

    private static func pythonGroup(cwd: String, names: Set<String>) -> ProjectCommandGroup {
        var commands: [ProjectCommand] = []

        if names.contains("requirements.txt") {
            commands.append(ProjectCommand(
                label: "pip install", command: "pip install -r requirements.txt",
                systemImage: "square.and.arrow.down"
            ))
        }
        if names.contains("pyproject.toml") || names.contains("setup.py") {
            commands.append(ProjectCommand(
                label: "pip install -e", command: "pip install -e .",
                systemImage: "square.and.arrow.down"
            ))
        }
        if names.contains("manage.py") {
            commands.append(ProjectCommand(
                label: "runserver", command: "python manage.py runserver",
                systemImage: "play.fill"
            ))
        } else if names.contains("app.py") {
            commands.append(ProjectCommand(
                label: "Run app.py", command: "python app.py", systemImage: "play.fill"
            ))
        } else if names.contains("main.py") {
            commands.append(ProjectCommand(
                label: "Run main.py", command: "python main.py", systemImage: "play.fill"
            ))
        }
        commands.append(ProjectCommand(
            label: "pytest", command: "pytest", systemImage: "checkmark.seal.fill"
        ))

        return ProjectCommandGroup(name: "Python", systemImage: "terminal", commands: commands)
    }

    private static func makeTargets(cwd: String) -> [ProjectCommand] {
        let makePath = (cwd as NSString).appendingPathComponent("Makefile")
        guard let content = try? String(contentsOfFile: makePath, encoding: .utf8) else {
            return [ProjectCommand(label: "make", command: "make", systemImage: "play.fill")]
        }

        var targets: [String] = []
        for line in content.components(separatedBy: .newlines) {
            guard !line.hasPrefix("\t"), !line.hasPrefix("#"), !line.hasPrefix(".") else { continue }
            if let colonIdx = line.firstIndex(of: ":") {
                let target = String(line[..<colonIdx]).trimmingCharacters(in: .whitespaces)
                if !target.isEmpty && !target.contains(" ") && !target.contains("=") {
                    targets.append(target)
                }
            }
        }

        let limited = targets.prefix(6)
        guard !limited.isEmpty else {
            return [ProjectCommand(label: "make", command: "make", systemImage: "play.fill")]
        }
        return limited.map {
            ProjectCommand(label: "make \($0)", command: "make \($0)", systemImage: scriptIcon($0))
        }
    }

    private static func scriptIcon(_ name: String) -> String {
        switch name {
        case "dev", "start", "serve", "preview", "runserver": return "play.fill"
        case "build":   return "hammer.fill"
        case "test":    return "checkmark.seal.fill"
        case "lint", "check", "format": return "exclamationmark.circle"
        case "clean":   return "trash"
        default:        return "terminal"
        }
    }
}

File: windows/app/FriendlyTerminal.App/Models/ProjectCommands.cs
using System.IO;
using System.Text.Json;

namespace FriendlyTerminal.App.Models;

public sealed record ProjectCommandGroup(string Name, int Glyph, IReadOnlyList<ProjectCommand> Commands);

public sealed record ProjectCommand(string Label, string Command, int Glyph);

/// <summary>
/// Suggests ready-to-run commands for the project in the current folder
/// (npm/pnpm/yarn scripts, Python, Rust, Go, Make, Ruby, Docker), mirroring the
/// macOS ProjectCommandDetector. Glyphs are Segoe Fluent Icons code points.
/// </summary>
public static class ProjectCommandDetector
{
    private const int Play = 0xE768;
    private const int Build = 0xE90F;
    private const int Check = 0xE73E;
    private const int Download = 0xE896;
    private const int TerminalIcon = 0xE756;
    private const int Warn = 0xE7BA;
    private const int Trash = 0xE74D;

    public static List<ProjectCommandGroup> Suggestions(string cwd, IEnumerable<FileEntry> files)
    {
        var names = new HashSet<string>(files.Select(f => f.Name), StringComparer.OrdinalIgnoreCase);
        var groups = new List<ProjectCommandGroup>();

        if (names.Contains("package.json") && NodeGroup(cwd, names) is { } node)
            groups.Add(node);

        if (names.Contains("requirements.txt") || names.Contains("pyproject.toml") || names.Contains("setup.py"))
            groups.Add(PythonGroup(names));

        if (names.Contains("Cargo.toml"))
        {
            groups.Add(new ProjectCommandGroup("Rust", 0xE713, new[]
            {
                new ProjectCommand("Run", "cargo run", Play),
                new ProjectCommand("Build", "cargo build", Build),
                new ProjectCommand("Test", "cargo test", Check),
            }));
        }

        if (names.Contains("go.mod"))
        {
            groups.Add(new ProjectCommandGroup("Go", 0xE943, new[]
            {
                new ProjectCommand("Run", "go run .", Play),
                new ProjectCommand("Test", "go test ./...", Check),
                new ProjectCommand("Build", "go build .", Build),
            }));
        }

        if (names.Contains("Makefile"))
            groups.Add(new ProjectCommandGroup("Make", Build, MakeTargets(cwd)));

        if (names.Contains("Gemfile"))
        {
            groups.Add(new ProjectCommandGroup("Ruby", 0xE7B8, new[]
            {
                new ProjectCommand("Bundle install", "bundle install", Download),
                new ProjectCommand("RSpec", "bundle exec rspec", Check),
            }));
        }

        if (names.Contains("Dockerfile"))
        {
            groups.Add(new ProjectCommandGroup("Docker", 0xE7B8, new[]
            {
                new ProjectCommand("Build", "docker build .", Build),
            }));
        }

        return groups;
    }

    private static ProjectCommandGroup? NodeGroup(string cwd, HashSet<string> names)
    {
        string manager;
        if (names.Contains("pnpm-lock.yaml") || names.Contains("pnpm-lock.yml")) manager = "pnpm";
        else if (names.Contains("yarn.lock")) manager = "yarn";
        else manager = "npm";

        var commands = new List<ProjectCommand>();

        if (!names.Contains("node_modules"))
            commands.Add(new ProjectCommand("Install deps", $"{manager} install", Download));

        var scripts = ReadPackageScripts(Path.Combine(cwd, "package.json"));
        if (scripts.Count > 0)
        {
            string[] priority = { "dev", "start", "serve", "preview", "build", "test", "lint", "check", "format" };
            foreach (var key in priority.Where(scripts.Contains))
            {
                var cmd = manager == "npm" ? $"npm run {key}" : $"{manager} {key}";
                commands.Add(new ProjectCommand(key, cmd, ScriptGlyph(key)));
            }
            foreach (var key in scripts.Where(s => !priority.Contains(s)).OrderBy(s => s).Take(4))
            {
                var cmd = manager == "npm" ? $"npm run {key}" : $"{manager} {key}";
                commands.Add(new ProjectCommand(key, cmd, TerminalIcon));
            }
        }
        else
        {
            commands.Add(new ProjectCommand("dev", manager == "npm" ? "npm run dev" : $"{manager} dev", Play));
            commands.Add(new ProjectCommand("test", manager == "npm" ? "npm test" : $"{manager} test", Check));
            commands.Add(new ProjectCommand("build", manager == "npm" ? "npm run build" : $"{manager} build", Build));
        }

        if (commands.Count == 0) return null;
        var displayName = manager switch { "pnpm" => "pnpm", "yarn" => "Yarn", _ => "Node.js" };
        return new ProjectCommandGroup(displayName, 0xE943, commands);
    }

    private static List<string> ReadPackageScripts(string packageJsonPath)
    {
        try
        {
            using var doc = JsonDocument.Parse(File.ReadAllText(packageJsonPath));
            if (doc.RootElement.TryGetProperty("scripts", out var scripts)
                && scripts.ValueKind == JsonValueKind.Object)
                return scripts.EnumerateObject().Select(p => p.Name).ToList();
        }
        catch { }
        return new List<string>();
    }

    private static ProjectCommandGroup PythonGroup(HashSet<string> names)
    {
        var commands = new List<ProjectCommand>();

        if (names.Contains("requirements.txt"))
            commands.Add(new ProjectCommand("pip install", "pip install -r requirements.txt", Download));
        if (names.Contains("pyproject.toml") || names.Contains("setup.py"))
            commands.Add(new ProjectCommand("pip install -e", "pip install -e .", Download));

        if (names.Contains("manage.py"))
            commands.Add(new ProjectCommand("runserver", "python manage.py runserver", Play));
        else if (names.Contains("app.py"))
            commands.Add(new ProjectCommand("Run app.py", "python app.py", Play));
        else if (names.Contains("main.py"))
            commands.Add(new ProjectCommand("Run main.py", "python main.py", Play));

        commands.Add(new ProjectCommand("pytest", "pytest", Check));
        return new ProjectCommandGroup("Python", TerminalIcon, commands);
    }

    private static IReadOnlyList<ProjectCommand> MakeTargets(string cwd)
    {
        string content;
        try { content = File.ReadAllText(Path.Combine(cwd, "Makefile")); }
        catch { return new[] { new ProjectCommand("make", "make", Play) }; }

        var targets = new List<string>();
        foreach (var line in content.Split('\n'))
        {
            if (line.StartsWith('\t') || line.StartsWith('#') || line.StartsWith('.')) continue;
            var colon = line.IndexOf(':');
            if (colon <= 0) continue;
            var target = line[..colon].Trim();
            if (target.Length > 0 && !target.Contains(' ') && !target.Contains('='))
                targets.Add(target);
        }

        if (targets.Count == 0)
            return new[] { new ProjectCommand("make", "make", Play) };
        return targets.Take(6)
            .Select(t => new ProjectCommand($"make {t}", $"make {t}", ScriptGlyph(t)))
            .ToList();
    }

    private static int ScriptGlyph(string name) => name switch
    {
        "dev" or "start" or "serve" or "preview" or "runserver" => Play,
        "build" => Build,
        "test" => Check,
        "lint" or "check" or "format" => Warn,
        "clean" => Trash,
        _ => TerminalIcon,
    };
}
